### PR TITLE
EntitlementVerification: Add verification property to EntitlementInfos and EntitlementInfo

### PR DESCRIPTION
**Pull request by tonidero**
### Description

Changes:
- Moved `HTTPResult.VerificationStatus` from the common module to `VerificationResult` in the public module.
- Add the property `verification` of type `VerificationResult` to `EntitlementInfos` and `EntitlementInfo`.
- Change `CustomerInfo` creation from `jsonObject.buildCustomerInfo()` to `CustomerInfoFactory.buildCustomerInfo(jsonObject, verificationResult)`.

### CODE REVIEW

File: public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt [hunk 1]
```diff
@@ -49,7 +49,8 @@ data class EntitlementInfo constructor(
     @Deprecated(
         "Use rawData instead",
         replaceWith = ReplaceWith("rawData")
-    ) val jsonObject: JSONObject
+    ) val jsonObject: JSONObject,
+    val verification: VerificationResult
```
Comment (tonidero) [inline public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt:53]: This and the change in `EntitlementInfos` are technically breaking changes, that's part of what I wanted to discuss over slack. In this case, I could default to `NOT_VERIFIED` but I don't think that's ideal.

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import com.revenuecat.purchases.utils.parsePurchaseDates
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.Collections.emptyMap
+
+/**
+ * Builds a CustomerInfo
+ * @throws [JSONException] If the json is invalid.
+ */
+object CustomerInfoFactory {
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt:20]: This change caused most of the other changes on this PR. I thought making the CustomerInfo construction be a part of JSONObject didn't make sense anymore when there is more data we need now. I thought about adding the verification result to the json, but that didn't make sense to me since that json is supposed to be the response body from the backend and I didn't want to mix data. I think this way is also more extensible in case creating a CustomerInfo changes in the future.

File: common/src/main/java/com/revenuecat/purchases/common/caching/DeviceCache.kt [hunk 3]
```diff
@@ -110,8 +122,9 @@ open class DeviceCache(
 
     @Synchronized
     fun cacheCustomerInfo(appUserID: String, info: CustomerInfo) {
-        val jsonObject = info.jsonObject.also {
-            it.put("schema_version", PURCHASER_INFO_SCHEMA_VERSION)
+        val jsonObject = info.rawData.also {
+            it.put(CUSTOMER_INFO_SCHEMA_VERSION_KEY, PURCHASER_INFO_SCHEMA_VERSION)
+            it.put(CUSTOMER_INFO_VERIFICATION_RESULT_KEY, info.entitlements.verification.name)
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/caching/DeviceCache.kt:127]: This is a not ideal consequence of caching `CustomerInfo` but having the verification result property live in `EntitlementInfos`. But I think it's not that bad. Lmk what you think.

File: common/src/main/java/com/revenuecat/purchases/common/caching/DeviceCache.kt [hunk 4]
```diff
@@ -96,9 +102,15 @@ open class DeviceCache(
             ?.let { json ->
                 try {
                     val cachedJSONObject = JSONObject(json)
-                    val schemaVersion = cachedJSONObject.optInt("schema_version")
+                    val schemaVersion = cachedJSONObject.optInt(CUSTOMER_INFO_SCHEMA_VERSION_KEY)
+                    val verificationResultString = if (cachedJSONObject.has(CUSTOMER_INFO_VERIFICATION_RESULT_KEY)) {
+                        cachedJSONObject.getString(CUSTOMER_INFO_VERIFICATION_RESULT_KEY)
+                    } else VerificationResult.NOT_VERIFIED.name
+                    cachedJSONObject.remove(CUSTOMER_INFO_SCHEMA_VERSION_KEY)
+                    cachedJSONObject.remove(CUSTOMER_INFO_VERIFICATION_RESULT_KEY)
```
Comment (tonidero) [inline common/src/main/java/com/revenuecat/purchases/common/caching/DeviceCache.kt:110]: Note how I'm removing the extra fields that are added during caching. This is to avoid having the underlying json have extra information it doesn't need when caching it. Before, we were adding the `schema_version` field to the underlying json after caching it, so it wasn't technically equal.

File: public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt [hunk 5]
```diff
@@ -49,9 +49,43 @@ data class EntitlementInfo constructor(
...
 ) : Parcelable, RawDataContainer<JSONObject> {
 
+    @Deprecated("Use full constructor instead")
```
Comment (tonidero) [inline public/src/main/java/com/revenuecat/purchases/EntitlementInfo.kt:56]: As we discussed over Slack, I'm adding a deprecated constructor with a default value for now. Reason to not simply add a default value to the field is that I don't think we should allow creating these models without this object, and adding a default value can lead to forgetting adding that field. This way at least we would get a deprecated warning.

File: api-tester/src/main/java/com/revenuecat/apitester/java/EntitlementInfosAPI.java [hunk 6]
```diff
@@ -2,14 +2,21 @@
...
+@SuppressWarnings({"unused", "deprecation"})
 final class EntitlementInfosAPI {
     static void check(final EntitlementInfos infos) {
         final Map<String, EntitlementInfo> active = infos.getActive();
         final Map<String, EntitlementInfo> all = infos.getAll();
         final EntitlementInfo i = infos.get("");
+        final VerificationResult verification = infos.getVerification();
+    }
+
+    static void checkConstructor(Map<String, EntitlementInfo> all, VerificationResult verificationResult) {
```
Comment (tonidero) [inline api-tester/src/main/java/com/revenuecat/apitester/java/EntitlementInfosAPI.java:18]: We should add these for all public models that have a public constructor, but we can do that in a different PR.

**Comment by tonidero** [comment]
Oh the new deprecated constructors are not covered by tests, that seems to be the reason coverage went down... Seems like it's not worth adding tests for that, since we are just keeping them for retrocompatibility and there is no logic but lmk if you think otherwise.

File: common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    @Throws(JSONException::class)
+    fun buildCustomerInfo(httpResult: HTTPResult): CustomerInfo {
+        return buildCustomerInfo(httpResult.body, httpResult.verificationResult)
```
Comment (NachoSoto) [inline common/src/main/java/com/revenuecat/purchases/common/CustomerInfoFactory.kt:23]: This is clean 👍🏻

File: common/src/test/java/com/revenuecat/purchases/common/MockCustomerInfoFactory.kt [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package com.revenuecat.purchases.common
+
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.VerificationResult
+import org.json.JSONObject
+
+fun createCustomerInfo(
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply common/src/test/java/com/revenuecat/purchases/common/MockCustomerInfoFactory.kt:7]: Yeah these types of helpers are great. Tests are "real" code too, so refactoring them is just as important 👏🏻

**Review by NachoSoto — APPROVED** [review]
